### ISSUE TITLE
New execution model multiple nodes issue

### ISSUE THREAD

**Issue by gaow** [vatlab/sos#1290]
I'm running some (what I think are) multi-node jobs; here is my `squeue` output:

```
          JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)
          62477761   broadwl M40_9270     gaow  R       2:38      1 midway2-0262
          62477762   broadwl M40_f5cc     gaow  R       2:38      1 midway2-0317
          62477763   broadwl M40_7fc4     gaow  R       2:38      1 midway2-0169
          62477764   broadwl M40_0c87     gaow  R       2:38      1 midway2-0170
          62477765   broadwl M40_315f     gaow  R       2:38      1 midway2-0373
          62477766   broadwl M40_c4df     gaow  R       2:38      1 midway2-0374
```

as you can see `NODES` are 1. And for the generated sbatch file:

```
sos status M40_f5cc -v4
```
```
#SBATCH --time=03:00:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --cpus-per-task=1
#SBATCH --mem-per-cpu=8G
```

However my configuration file template is:

```
      #SBATCH --time={walltime}
      #SBATCH --nodes={nodes}
      #SBATCH --ntasks-per-node={cores}
      #SBATCH --cpus-per-task=1
      #SBATCH --mem-per-cpu={mem//10**9}G
      #SBATCH --job-name={job_name}
```

and my `task` statement is:

```
task: trunk_workers = [4, 10], trunk_size = 40, walltime = '20m', mem = '2G', cores = 1
```

that is, I ask for 10 nodes per job and on each node there will be 4 parallel processes. Then with trunk_size = 40 all jobs should be processed at once, not sequencially. Is that what supposed to happen? Then why do I still see `NODES` column has 1 for every job? I run it from login node directly via `sos run` not submitting any scripts to run it. But I assume either way will work. What am I missing here?

**Comment by BoPeng** [comment]
10 nodes should be [4]*10.

**Comment by gaow** [comment]
Oh right, I changed it to `[4]*10`, but I still see exactly the same -- one node used as shown in `squeue` and `sos status -v4`. I'm on the latest release 0.20.1. The pipeline I use is `[extract]` step of  : 

https://github.com/chj1ar/fine-mapping/blob/master/Multi_tissues.ipynb

**Comment by BoPeng** [comment]
Before checking anything else, could you add to you template

```
export SOS_DEBUG=WORKER
```
and report back the output of [this line of code](https://github.com/vatlab/sos/blob/master/src/sos/utils.py#L1904)?

**Comment by gaow** [comment]
Oh right I didnt use that in my current run. Okay let me terminate my current job, wait for those in the queue to complete (in a few hours from now) and rerun with that bash varaible set, then report back here.

**Comment by BoPeng** [comment]
You can create a test job to see if the parameter is handled correctly.

**Comment by gaow** [comment]
Already did, but the queue is saturated so I'll see if after a few hours it works out.

**Comment by gaow** [comment]
Okay I terminated my other jobs because I"ll have to leave in 20min. Now the test job completed and I get: 

```
2019-09-15 17:56:09,780 - DEBUG - WORKER - Using "-j ['midway2-0021:1']" on a SLURM cluster.
2019-09-15 17:56:09,797 - DEBUG - WORKER - Using "-j ['midway2-0016:1']" on a SLURM cluster.
```

The test job is:

```
[1]
genes = [x for x in range(80)]
input: for_each = 'genes'
output: f'{_genes}.txt'
task: trunk_workers = [4]*10, trunk_size = 40, walltime = '1m', mem = '1G', cores = 1
bash: expand = True
  touch {_output}
```

and the command: 

```
sos run test.sos -c midway2.yml -q midway2
```

the config:

https://github.com/chj1ar/fine-mapping/blob/master/midway2.yml

**Comment by BoPeng** [comment]
I cannot log in so you will have to debug this. Basically you will need to print the value of `SLURM_NODELIST` and output of command `scontrol show hostnames {SLURM_NODELIST}`.

**Comment by gaow** [comment]
As discussed offline, now after updating `sos-pbs` it seems to be working as expected:

```
         62481896   broadwl M40_b24b     gaow PD       0:00     10 (QOSMaxJobsPerUserLimit)
         62482423   broadwl M40_b24b     gaow PD       0:00     10 (QOSMaxJobsPerUserLimit)
```

showing 10 nodes are being used. However the job template still looks like:

```
#SBATCH --partition=broadwl
#
#SBATCH --time=00:09:00
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=4
#SBATCH --cpus-per-task=1
...
```
(using command `sos status M40_55ea003b7aa47fdc -v4`)

Is this expected? In other words is the openmpi part hidden behind the scene now?

**Comment by BoPeng** [comment]
Are you sure you are checking the right `.sh` file? you should see `--nodes=10` as what I saw.

**Comment by BoPeng** [comment]
Sorry, `sos status M40_55ea003b7aa47fdc` might be reporting the old `.sh` file now since the task has not been finished and the `.sh` file has not been "absorbed". I was referring to the physical file `~/.sos/tasks/M40_55ea003b7aa47fdc.sh` when the task is running.

**Comment by gaow** [comment]
Great! Yes it worked great, I am able to use it in my production code now:

```
          62484156   broadwl M40_e7be     gaow PD       0:00     10 (Resources)
          62484157   broadwl M40_2188     gaow PD       0:00     10 (Resources)
          62484158   broadwl M40_d6fe     gaow PD       0:00     10 (Resources)
          62484159   broadwl M40_3621     gaow PD       0:00     10 (Resources)
          62484160   broadwl M40_3b24     gaow PD       0:00     10 (Resources)
          62484161   broadwl M40_47c9     gaow PD       0:00     10 (Resources)
          62484162   broadwl M40_fef0     gaow PD       0:00     10 (Resources)
          62484167   broadwl M40_16c5     gaow PD       0:00     10 (Resources)
          62484170   broadwl M40_5b50     gaow PD       0:00     10 (Resources)
          62484171   broadwl M40_f0e2     gaow PD       0:00     10 (Resources)
          62484172   broadwl M40_a78e     gaow PD       0:00     10 (None)
          62484173   broadwl M40_46c4     gaow PD       0:00     10 (None)
          62484174   broadwl M40_ba40     gaow PD       0:00     10 (None)
          62476713   broadwl nb-tunne     gaow  R   19:42:50      1 midway2-0010
          62484136   broadwl M40_742a     gaow  R       0:06     10 midway2-[0029,0047,0052,0059,0337-0338,0361-0362,0397-0398]
          62484137   broadwl M40_4196     gaow  R       0:06     10 midway2-[0061,0063,0065,0071,0073,0112,0118,0139,0154,0170]
          62484135   broadwl M40_9a9b     gaow  R       0:08     10 midway2-[0003,0019-0020,0055-0056,0078-0079,0383-0385]
          62484127   broadwl M40_1e81     gaow  R       0:25     10 midway2-[0176,0224,0227,0260,0263,0265,0267,0280,0284,0311]
[MW] 
```

However, still setting my max allowed jobs to 150 is now a bit daunting, as you can see many of them are in the queue. I'll adjust my max allowed jobs for this case to a reasonable number.

**Comment by gaow** [comment]
BTW I still receive this warning from time to time: 

```
WARNING: Task M40_6bc3e34572ba15e5 considered as aborted due to missing pulse file.
```

**Comment by BoPeng** [comment]
Checking the header of this task file,

```
>>> a.info
TaskHeader(version=3, status=4, 
last_modified=1568643964.2333164, 
new_time=1568584309.5246372, 
pending_time=1568643715.4627838, 
submitted_time=1568643726.018421, 
running_time=1568643963.8671649, 
aborted_time=1568643964.2333164, 
failed_time=1568604885.2664855, 
completed_time=0.0, 
params_size=20248, 
runtime_size=316, shell_size=0, pulse_size=0, stdout_size=0, stderr_size=132, result_size=0, signature_size=0, tags=b'bea9d36b1b500b8c extract extract_ENSG00000165028.11.Multi_Tissues extract_ENSG00000165029.15.Multi_Tissues extract_ENSG0')
```
There is a failed time after new time, and before submitted time, which means the task has been re-submitted (but `failed_time` should perhaps be reset). It waited quite a bit, set to `running` and very soon found to be `aborted` due to lack of `pulse` file.

According to source code [here](https://github.com/vatlab/sos/blob/master/src/sos/tasks.py#L694), the `pulse` file is created before the status is set to `running`, and it actually checks the availability of the `pulse` file before setting status of task file. Not quite sure what is going on.

**Comment by gaow** [comment]
@BoPeng sorry I am still having issues with the template, 

```
      #SBATCH --time={walltime}
      #SBATCH --nodes={nodes}
      #SBATCH --cpus-per-task={cores}
      #SBATCH --mem={mem//10**9}G
```

and in my `task:`

```
 walltime = '5m', mem = '2G', cores = 1, trunk_size = 120, trunk_workers = [4] * 10, 
```

However in `sos status` I see:

```
#SBATCH --time=02:25:00
#SBATCH --nodes=10
#SBATCH --cpus-per-task=4
#SBATCH --mem=8G
```

I get that cores is 4 and memory is 2*4=8GB due to trunk worker of 4. However for trunk_size of 120, if I use 10 nodes, each node uses 4 cores, then I should have 120 / 10 / 4 = 3 sequential jobs which should be 15min wall time. `2:25` is 145 minutes which is 29*5 ... not sure how it got that number.

**Comment by BoPeng** [comment]
I see, `walltime` does not yet consider multiple nodes now.

**Comment by gaow** [comment]
Thanks! I guess that then warrants another `sos-pbs ` release? :P

**Comment by BoPeng** [comment]
Try again? It would be a sos release because the planning is done at the sos side.

**Comment by gaow** [comment]
Great -- now it works as expected! 

However before releasing a new version: I think it takes a lot longer to prepare for large number tasks than past versions. I'll see if I can show this by one of our stress tests and report it in that repo. Hopefully our next release can also fix that.

I'll close the ticket as both bugs seem to have been fixed.

**Comment by gaow** [comment]
@BoPeng sorry let me reuse this ticket to report another issue. My script looks like:

```
[1]
genes = [x for x in range(100)]
input: for_each = 'genes'
task: trunk_size=5, trunk_workers = [4]*1, mem = '1G', walltime='5m',queue='midway2'
R: expand = True
  while(1) x = {_genes}
```

and my job template: 


```

      #SBATCH --time={walltime}
      #SBATCH --nodes={nodes}
      #SBATCH --cpus-per-task={cores}
```

so I would assume `{cores}` is 4. However, for the job I submitted:

```
$ sos status M5_de77f -v4

#SBATCH --time=00:05:00
#SBATCH --nodes=1
#SBATCH --cpus-per-task=1
#SBATCH --mem=4G
```

you see CPUs per task is 1, not 4 as I would expect. What am I missing here?

**Comment by gaow** [comment]
Also the walltime doesnt seem correct: for trunk_size = 5 but asking for 4 trunk workers, the wall time should be 10min: 4 in parallel for a total 5 min, then the other one by itself and take another 5min?

**Comment by gaow** [comment]
Oh I want to add that I'm on current `master` and I have removed all `~/.sos` folder. So this seems a regression bug ..

**Comment by BoPeng** [comment]
I think this is caused by a bug with no `cores` setting. Can you try with `cores=1`?

**Comment by gaow** [comment]
hmm, with `cores=1` I get:

```
#SBATCH --time=00:05:00
#SBATCH --nodes=1
#SBATCH --cpus-per-task=4
#SBATCH --mem=4G
```

though the walltime still doesnt look right?

**Comment by gaow** [comment]
Also, when I `ssh ` to the actual compute node I see this with `top -u gaow`

```
55487 gaow      20   0  161316  48880   6276 R 100.0  0.1   1:55.53 R                     
55546 gaow      20   0  161312  48876   6276 R  99.7  0.1   1:54.30 R                     
56349 gaow      20   0   58868   2780   1632 R   0.7  0.0   0:00.25 top                   
55451 gaow      20   0   11780   1616   1272 S   0.0  0.0   0:00.00 slurm_script          
55473 gaow      20   0  172832  27440   6952 S   0.0  0.0   0:00.74 sos                   
55517 gaow      20   0   11780   1620   1272 S   0.0  0.0   0:00.00 slurm_script          
55540 gaow      20   0  172832  27444   6952 S   0.0  0.0   0:00.64 sos 
```

You see here there are 2 R processes in parallel, not 4. Perhaps this is because the other 2 processes are used by `sos`? But in this case the user would expect 4 threads for the computation alone. ..

**Comment by BoPeng** [comment]
OK, `walltime=5m`, trunksize=5, with 4 workers, so there are two batches, and walltime should be 10m. Should be some sort of integer division problem.

**Comment by BoPeng** [comment]
Try again?

**Comment by gaow** [comment]
@BoPeng the code now seems to run, but when I log to individual compute nodes, eg, 

```
ssh midway2-0338
```

and 

```
top -u gaow
```

I only see one `R` instance running not 4. So it does not seem like SoS has used all the resource it was allocated to?

**Comment by BoPeng** [comment]
Right now if one worker is fast enough to respond all the requests, other workers would not be started. Remote workers are started all at the same time though.

Use `export SOS_DEBUG=WORKER` to check the behavior.

**Comment by gaow** [comment]
I'm not sure if I get your point. I should still expect to see 4 actual R instances right? I take you are referring to `sos` process as workers, but whatever it does, there should be a way to run multiple threads on a compute node because of `trunk_workers`. 

Here is my `sos_debug` file

[sos_debug.log](https://github.com/vatlab/sos/files/3633878/sos_debug.log)

**Comment by BoPeng** [comment]
No time to look in details now but basically sos starts one worker, and only when there is no "available" worker for new request will it starts another worker

https://github.com/vatlab/sos/blob/master/src/sos/workers.py#L521

If each request would start a new R process, perhaps the R processes die very quickly so that you only see one.

**Comment by gaow** [comment]
No worries. in that test (path and command sent to you via email) the R procedure is an infinite loop so it is not going to die until time out. Let me try fiddling with the template just in case it is my SLURM mis-specification.

**Comment by BoPeng** [comment]
Well, another unhandled case, the list `trunk_workers` but with only one node. Embarrassing but fixed now.

**Comment by BoPeng** [comment]
Made a new sos release.

**Comment by gaow** [comment]
Thank you for the prompt fix! It works as expected now. But the CI build seems to fail on Windows. Hopefully it is not important.

**Comment by BoPeng** [comment]
I restarted appveyor test and it passes now. I could have waited for #1264 though.

**Comment by gaow** [comment]
@BoPeng the multinode execution still doesnt quite work ... here is the MWE:

[test.tar.gz](https://github.com/vatlab/sos/files/3659264/test.tar.gz)

Here I ask 2 nodes per job. when I submitted and I ssh to one of the nodes I still see 1 instance of R instead of 4. However if I change the `trunk_workers` to `[4] * 1` it works as expected, due to a fix from above.

**Comment by BoPeng** [comment]
Your example has `trunk_size=5`, meaning there are at most 5 concurrent subtasks. When you have `trunk_workers=[4]*2`, all 8 workers will be created but it is very likely that only 1 will be sent to the remote worker, leading to one instance of `R`. Could you try again with `trunk_size=10` or more?

**Comment by gaow** [comment]
Sorry my cluster queue is now saturated so I cannot test. But in my production code where I noticed the issue I've done with `trunk_size=100, trunk_workers = [28] * 2` and still saw just one instance of R.

**Comment by BoPeng** [comment]
I cannot reproduce your problem here:

1. Verified there is no process on two nodes.
2. with script
```
$ cat test.sos
[1]
genes = [x for x in range(100)]
input: for_each = 'genes'
task: trunk_size=None, trunk_workers = [5]*2, mem = '1G', walltime='5m'
R: expand = True
  while(1) x = {_genes}
```
3. Find the nodes used with command `qstat -f `
4. The first node has 6 `sos execute` command, with 1 controller and 5 workers, and 5 R processes. It also has a `ssh` command to remote host.
5. The second node has 6 `sos execute`  processes and 5 R processes.

I did notice that `sos kill` will kill the `sos execute` process on the first node, but not on the second node. #1302

**Comment by gaow** [comment]
@BoPeng sorry i did not mean to send u on a wild goose chase. Let me confirm tmr on my slurm when the cluster queue is cleared up ...

**Comment by gaow** [comment]
@BoPeng maybe one issue at a time (RE #1302) but in my SLURM test things still not working:

Under `~/tmp/test` I've got that `test.sos` and as you suggested I put in larger trunk size:
```
task: trunk_size=20, trunk_workers = [4]*2, mem = '1G', walltime='5m',queue='midway2'
```

and my job looks like:

![Screenshot from 2019-09-30 08-44-14](https://user-images.githubusercontent.com/917985/65885028-7cff1180-e35f-11e9-80c9-0f868b8f7c8f.png)


Then I seeh to a particular node `midway2-0382` and see the following:

![Screenshot from 2019-09-30 08-42-26](https://user-images.githubusercontent.com/917985/65885090-999b4980-e35f-11e9-8519-7db6302e8a7a.png)


What is more disturbing is when I ssh to node `midway2-0383` also as listed above, I dont see any `R` instances! It would be great if you could reproduce there when you get a chance. Thanks!

**Comment by BoPeng** [comment]
OK, I see

```
"-j ['midway2-0371:1', 'midway2-0372:1']" on a SLURM cluster.
```
so this is a SLURM environment translation problem. Let me check.

**Comment by BoPeng** [comment]
`SLURM_NODELIST=midway2-[0178-0179]` so it does not contain process information. SoS currently assumes 1.

**Comment by gaow** [comment]
Is it a problem with translation, or job template? If I use `*1` for the node it works I believe. And why in the example above `midway2-0383` doesnt have any R instances?

**Comment by BoPeng** [comment]
Another variable `SLURM_JOB_CPUS_PER_NODE` is needed in this case and I have just submitted a patch. My original implementation coped [scoop's implementation](https://github.com/soravux/scoop/blob/master/scoop/utils.py#L204) and is incorrect.

However, there is a big issue of **jobs per node** (sos workers) and **cpus per task** (cores) and I am not sure if SoS is handling them correctly.

**Comment by gaow** [comment]
Thank you!

>  there is a big issue of jobs per node (sos workers) and cpus per task (cores

what issue? number of SoS workers and cores per worker doesnt seem conflicting. I think you mean "tasks" per node -- specific for SLURM.  Indeed `ntask` is what causes a confusion and is what I was contacting you about in some email threads. I think with SoS we will not need the concept of `ntask`. In the job template we simply dont set the number of tasks per node so that we use the default of 1.

**Comment by BoPeng** [comment]
>  I think with SoS we will not need the concept of ntask

Your understanding is correct, I have expanded it in #1304 and is double-checking is the logics are correct in the current implementation. You can also check if the implementation is correct by setting `cores` to more than `1`.

**Comment by gaow** [comment]
Now that I login and count the number of cores used I see it seems correct with the template I currently use.

**Comment by gaow** [comment]
Okay I think we can finally close this issue? I keep my fingers crossed that we dont have to reopen it. No rush, but looking forward to the next release!

**Comment by BoPeng** [comment]
I had to rely on your to check the behavior on SLURM, and hopefully everything works. Note that #1302 is still not fixed so kill tasks will leave zombie processes on computing nodes. 

I have double checked PBS, let me check a LSF system because make a release. We will have to rely on a Sun Grid user to make sure sos works correctly over there.

**Comment by BoPeng** [comment]
sos 0.20.6 released with the SLURM fix.
